pandas/io/sql.py

```diff
@@ -2024,6 +2042,255 @@ def _create_sql_schema(
 
 
 # ---- SQL without SQLAlchemy ---
+
+
+class ADBCDatabase(PandasSQL):
+    """
+    This class enables conversion between DataFrame and SQL databases
+    using ADBC to handle DataBase abstraction.
+
+    Parameters
+    ----------
+    con : adbc_driver_manager.dbapi.Connection
+    """
+
+    def __init__(self, con) -> None:
+        self.con = con
+
+    def execute(self, sql: str | Select | TextClause, params=None):
+        with self.con.cursor() as cur:
+            return cur.execute(sql)
+
+    def read_table(
+        self,
+        table_name: str,
+        index_col: str | list[str] | None = None,
+        coerce_float: bool = True,
+        parse_dates=None,
+        columns=None,
+        schema: str | None = None,
+        chunksize: int | None = None,
+        dtype_backend: DtypeBackend | Literal["numpy"] = "numpy",
+    ) -> DataFrame | Iterator[DataFrame]:
+        """
+        Read SQL database table into a DataFrame. Only keyword arguments used
+        are table_name and schema. The rest are silently discarded.
+
+        Parameters
+        ----------
+        table_name : str
+            Name of SQL table in database.
+        schema : string, default None
+            Name of SQL schema in database to read from
+
+        Returns
+        -------
+        DataFrame
+
+        See Also
+        --------
+        pandas.read_sql_table
+        SQLDatabase.read_query
+
+        """
+        if index_col:
+            raise NotImplementedError("'index_col' is not implemented for ADBC drivers")
+        if coerce_float is not True:
+            raise NotImplementedError(
+                "'coerce_float' is not implemented for ADBC drivers"
+            )
+        if parse_dates:
+            raise NotImplementedError(
+                "'parse_dates' is not implemented for ADBC drivers"
+            )
+        if columns:
+            raise NotImplementedError("'columns' is not implemented for ADBC drivers")
+        if chunksize:
+            raise NotImplementedError("'chunksize' is not implemented for ADBC drivers")
```

Comment:
Its a good question. I'm not sure I see anything in the ADBC specification around batch / chunk handling. Might be overlooking the general approach to that. @lidavidm always knows best

Reply:
Not necessarily in ABDC itself, but you should be able to get a RecordBatchReader instead of a materialized Table. And then pyarrow provides APIs to consume that reader chunk by chunk. It might not exactly support the user-specified chunksize, but it does give you a similar result: a generator of pandas DataFrames.

(a RecordBatchReader allows you to `read_next_batch()` at a time, this returns a RecordBatch, and then this could be further split if necessary based on `chunksize`, and then those chunks can be converted to pandas.DataFrame. The main issue is that if the native batch size you get from the database is much larger than the specified `chunksize`, you get a larger memory usage than expected)

Reply:
Oops, I missed this.

Drivers have various parameters to request a batch size but perhaps we should standardize on one.

Reply:
https://github.com/apache/arrow-adbc/issues/1024